controllers/topology/desired_state.go

```diff
@@ -170,6 +182,106 @@ func computeCluster(current *clusterTopologyState, infrastructureCluster, contro
 	return cluster
 }
 
+// computeMachineDeployment computes the desired state for a MachineDeploymentTopology.
+// The generated machineDeployment object is calculated using the values from the machineDeploymentTopology and
+// the machineDeployment class.
+func computeMachineDeployment(class *clusterTopologyClass, current *clusterTopologyState, topology clusterv1.MachineDeploymentTopology) (machineDeploymentTopologyState, error) {
+	desiredMachineDeployment := machineDeploymentTopologyState{}
+
+	className := topology.Class
+	machineDeploymentClass, err := getMachineDeploymentClass(class, className)
+	if err != nil {
+		return desiredMachineDeployment, err
+	}
+
+	machineDeploymentTopologyClass, ok := class.machineDeployments[className]
+	if !ok {
+		return desiredMachineDeployment, errors.Errorf("class name %s not found in clusterClass %s", className, class.clusterClass.Name)
+	}
+
+	currentMachineDeployment := current.machineDeployments[topology.Name]
+	var currentBootstrapTemplate *corev1.ObjectReference
+	if currentMachineDeployment.bootstrapTemplate != nil {
+		currentBootstrapTemplate = objToRef(currentMachineDeployment.bootstrapTemplate)
+	}
+	desiredMachineDeployment.bootstrapTemplate = templateToTemplate(templateToInput{
+		template:              machineDeploymentTopologyClass.bootstrapTemplate,
+		templateClonedFromRef: objToRef(machineDeploymentTopologyClass.bootstrapTemplate),
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-%s-bootstrap-", current.cluster.Name, topology.Name),
+		currentObjectRef:      currentBootstrapTemplate,
+		labels:                mergeMap(topology.Metadata.Labels, machineDeploymentClass.Template.Metadata.Labels),
+		annotations:           mergeMap(topology.Metadata.Annotations, machineDeploymentClass.Template.Metadata.Annotations),
+	})
+
+	var currentInfraMachineTemplate *corev1.ObjectReference
+	if currentMachineDeployment.infrastructureMachineTemplate != nil {
+		currentInfraMachineTemplate = objToRef(currentMachineDeployment.infrastructureMachineTemplate)
```

Comment:
Lets use the value from `currentMachineDeployment.object.spec.template.spec.infrastructureRef` which is the actual current ref (instead of relying on the object and the reference being consistent, which should be true, but it is a proxy information)